torch/nested/_internal/utils.py

```diff
@@ -0,0 +1,58 @@
+from typing import *  # noqa: F403
+
+import torch
+
+
+# Utility for nested subclass recursion with special handling for CachedTensor
+def apply_func(
```

Comment:
for this utility, it feels like there should be a more general way to recurse over traceable wrapper subclasses and their flattened forms (also taking into account any traceable wrapper subclasses that may exist in the flattened set).

`CacheTensor`'s flattened form should be able to take advantage of the above.

am I missing something? It'd be nice to not need specific NJT + CacheTensor logic here if we can just use the traceable wrapper subclass concept features

Reply:
It seemed difficult to me because we need NestedTensor specific logic here, in particular, I don't want to iterate through a CachedTensor field if its key is not a NestedTensor cache source key. CachedTensor itself has no concept of source vs non-source keys.

Another different is that the handling of subclasses that are not CachedTensor, what differs here with generic subclass recursion is that we are explicitly erroring. What we want to do here concerns how nesting subclasses works with NT wrt nested int comparison. Without thinking about that more, I'd like the default behavior to be erroring to avoid silently doing something weird. 

I can imagine something we might want if we try to make this more generic is to allow the generic subclass recursion code to also optionally accept some kind of "__tensor_unflatten__" override, where the user can say, given subclass A, instead of using its default tensor unflatten, I want to use my new tensor unflatten. I may also need to provide a default/catch all option, e.g. for remaining subclasses, I will have some kind of different default behavior.

TLDR: I think its possible, but maybe tricky, and maybe not worth it given that subclass recursion logic is simple enough and we're only using this once. But that being said, I agree that things are not the best and open to ideas that would make things cleaner.

Reply:
> in particular, I don't want to iterate through a CachedTensor field if its key is not a NestedTensor cache source key

this part you could easily cover with a "filter" arg indicating what to iterate over (default iterate over everything), could be a filter_fn predicate thing. There's no need to bring source vs. non-source keys into the iteration; that concept can be kept NJT-side where it belongs via the filter function.

> Another different is that the handling of subclasses that are not CachedTensor, what differs here with generic subclass recursion is that we are explicitly erroring.

when would this be encountered? It feels like whether erroring is desired would be dependent on the intention of the caller of this helper. To me, that's extra coupling between caller and helper that makes this hard to reason about.

In general, my goal throughout is to simplify as much as possible to be non-NJT-specific to make my life easier maintaining all this :)

Reply:
filter_fn sounds good to me!